### ISSUE TITLE
Injecting sidecar breaks websocket connection

### ISSUE THREAD

**Issue by faxg** [istio/istio#824]
Deploying Node-red on k8s and exposing through ingress works fine until I modify my deployment to inject istio sidecar.  

```kubectl describe ing```
-->
```
Name:                   hello-nodered
Namespace:              default
Address:
Default backend:        default-http-backend:80 (11.1.25.132:8080)
Rules:
  Host  Path    Backends
  ----  ----    --------

  *
        /hello  hello-nodered:1880 (<none>)
Annotations:
  websocket-services:   hello-nodered
  enable-cors:          true
  rewrite-target:       /
  ssl-redirect:         false
No events.
````
Now, if I do 
````
kubectl get deployment/hello-nodered -o yaml | istioctl kube-inject -f - | kubectl apply -f -
````
the websocket connection from the editor does not work anymore:
![image](https://user-images.githubusercontent.com/3209819/30498406-efe6c138-9a56-11e7-977d-f12b70b01139.png)

I tried both ingress annotations ```kubernetes.io/ingress.class: "istio"``` and ```kubernetes.io/ingress.class: "nginx"```, but always same issue. So I really suspect this is caused by the injected sidecar proxy.

**Comment by kyessenov** [comment]
cc @rshriram as a websocket expert

**Comment by rshriram** [comment]
You need to create a route rule for the service represented by Node-Red.. and in that route rule, you need to specify that the connections should be upgraded to WebSockets..
```yaml
metadata:
  name: my-rule
  namespace: default
spec:
  destination:
    name: ratings
  match:
    request:
        uri:
          prefix: "/ratings/v2/"
    route:
       - labels:
            version: 1
         weight: 100
    websocketUpgrade: true
```

**Comment by hollinwilkins** [comment]
I am creating the route, but now I get a websocket that just hangs and never sends/receives data.

Istio version is 0.3.0

Here is my route:

```
apiVersion: config.istio.io/v1alpha2
kind: RouteRule
metadata:
  name: ws-route
spec:
  websocketUpgrade: True
  destination:
    name: ws-service
  match:
    request:
      headers:
        uri:
          prefix: /api/conversation-events
```

**Comment by hollinwilkins** [comment]
@rshriram Would love to get your help solving this issue, I can get you any information you need to help debug, just not sure where to start.

**Comment by rshriram** [comment]
@hollinwilkins thanks for the report. I need more information on this. Possibly the Envoy configuration from both the source and destination pod to make sure the websocket configuration has propagated properly.
You can get the configuration from the local envoyusing `curl http://localhost:15000/routes` (its either 15000 or 15001). You should see a `use_websocket: true` on both configs.

**Comment by hollinwilkins** [comment]
@rshriram Will be happy to. Need to wait for a time when I can experiment in the cluster, unfortunately injecting sidecars also breaks out DB connections to postgres, which I am still figuring out. As a result, I don't always has istio injected, and it makes testing things like this slower than I would like.

**Comment by hollinwilkins** [comment]
@rshriram  `curl http://localhost:15000/routes |grep websocket` returns nothing.

**Comment by hollinwilkins** [comment]
@rshriram I see this:
```
"routes":[{"match":{"prefix":"/api/conversation-events"},"route":{"cluster":"out.ws-service.default.svc.cluster.local|http","timeout":"0s","use_websocket":true},"decorator":{"operation":"ws-route"}}
```

And on the ingress controller:

```
{"cluster":"out.ws-service.default.svc.cluster.local|http","timeout":"0s","use_websocket":true}
```

Looks like the `use_websocket` flag is propagating properly. In chrome, I get status pending for exactly 1 minute, and then the WS status goes to finished.

Worth noting:
1. The websocket is routed through an instance of nginx. There is no problem with this setup until I inject the istio sidecar, but maybe nginx websocket proxy does not play nice with envoy/istio for some reason
2. Nginx is proxying the websocket request to a localhost port with the websocket service running on it

**Comment by rshriram** [comment]
I need the envoy config (the /routes) from the ws-service.. What you showed above is for the outbound path (and hence the out.xxxxx cluster name). When you query ws-service's envoy, you should see a in.abcde... cluster with a use_websocket: true for the prefix you defined.

**Comment by hollinwilkins** [comment]
`curl http://localhost:15000/routes |grep "in\."` returns nothing from the istio-proxy of my websocket service

`curl http://localhost:15000/routes |grep "in\."` returns nothing from the ingress controller either

**Comment by rshriram** [comment]
If you list all the clusters (/clusters) , do you see any cluster with in.... or inbound.. ?
If there is no in.., then that indicates the issue. The inbound cluster for the service is not being created, and there is no route to receive the websocket in Envoy.

**Comment by hollinwilkins** [comment]
In clusters on the WS service:

```
in.80::default_priority::max_connections::1024
in.80::default_priority::max_pending_requests::1024
in.80::default_priority::max_requests::1024
in.80::default_priority::max_retries::3
in.80::high_priority::max_connections::1024
in.80::high_priority::max_pending_requests::1024
in.80::high_priority::max_requests::1024
in.80::high_priority::max_retries::3
in.80::added_via_api::true
in.80::127.0.0.1:80::cx_active::0
in.80::127.0.0.1:80::cx_connect_fail::0
in.80::127.0.0.1:80::cx_total::0
in.80::127.0.0.1:80::rq_active::0
in.80::127.0.0.1:80::rq_error::0
in.80::127.0.0.1:80::rq_success::0
in.80::127.0.0.1:80::rq_timeout::0
in.80::127.0.0.1:80::rq_total::0
in.80::127.0.0.1:80::health_flags::healthy
in.80::127.0.0.1:80::weight::1
in.80::127.0.0.1:80::region::
in.80::127.0.0.1:80::zone::
in.80::127.0.0.1:80::sub_zone::
in.80::127.0.0.1:80::canary::false
in.80::127.0.0.1:80::success_rate::-1
```

And

```
in.5877::default_priority::max_connections::1024
in.5877::default_priority::max_pending_requests::1024
in.5877::default_priority::max_requests::1024
in.5877::default_priority::max_retries::3
in.5877::high_priority::max_connections::1024
in.5877::high_priority::max_pending_requests::1024
in.5877::high_priority::max_requests::1024
in.5877::high_priority::max_retries::3
in.5877::added_via_api::true
in.5877::127.0.0.1:5877::cx_active::0
in.5877::127.0.0.1:5877::cx_connect_fail::0
in.5877::127.0.0.1:5877::cx_total::12
in.5877::127.0.0.1:5877::rq_active::0
in.5877::127.0.0.1:5877::rq_error::0
in.5877::127.0.0.1:5877::rq_success::0
in.5877::127.0.0.1:5877::rq_timeout::0
in.5877::127.0.0.1:5877::rq_total::0
in.5877::127.0.0.1:5877::health_flags::healthy
in.5877::127.0.0.1:5877::weight::1
in.5877::127.0.0.1:5877::region::
in.5877::127.0.0.1:5877::zone::
in.5877::127.0.0.1:5877::sub_zone::
in.5877::127.0.0.1:5877::canary::false
in.5877::127.0.0.1:5877::success_rate::-1
```

**Comment by rshriram** [comment]
Do you have any in.80 in /routes ? and if so,whats the configuration for it?

**Comment by hollinwilkins** [comment]
```curl http://localhost:15000/routes |grep "in\.80"``` returns nothing

**Comment by hollinwilkins** [comment]
Could it be because I don't have a container port specified for port 80?

**Comment by hollinwilkins** [comment]
```
apiVersion: v1
kind: Service
metadata:
  labels: ....
spec:
  ports:
  - name: http
    protocol: TCP
    port: 80
    targetPort: 80
  selector: ....
```

**Comment by hollinwilkins** [comment]
That's not the issue, I setup my container with a port named `http` and a containerPort set to `80` and I still have no `in.80` rules coming back from envoy.

**Comment by hollinwilkins** [comment]
@rshriram Is there any other information I can collect to help figure this out?

**Comment by ZihanLi0920** [comment]
I just meet the same problem. It will be really helpful if you could tell me how did you solve it.

**Comment by ymesika** [comment]
@ZihanLi0920 If you are trying to enable websockets on v0.8.0 with v1alpha3 rule please notice that a fix got merged only recently (#6073).

If it's an earlier version then it should work. I have a [v0.7.1/v1alpha2 demo app](https://gist.github.com/ymesika/c64326e3d8b4ff2c1e2b53c1e13ac72a) that's working for me. You can use it as a reference.

**Comment by ymesika** [comment]
Just tested the [sample Tornado app](https://gist.github.com/ymesika/c64326e3d8b4ff2c1e2b53c1e13ac72a) with v1alpha3 + sidecar injected and websocket seems to work fine with `master`.
Will close this one but feel free to re-open if you encounter problems.
